DNS: threaded resolver [WIP]

ysbaddaden opened the pull request:
This Pull Request proposes to call `getaddrinfo` within threads and to synchronize with the main thread to receive the request and send back the response, thus allowing to not stop-the-world whenever a Fiber needs to resolve a Domain Name. The threadpool size can be adjusted with `Socket::Addrinfo.threadpool_size = 4` for example before a Domain Name is first resolved, or even disabled with `Socket::Addrinfo.threadpool_size = 0`.
- [x] start configurable threadpool;
- [x] call `getaddrinfo` in threads;
- [x] fallback to call `getaddrinfo` from the main thread if the threadpool is disabled (ie. to set `0`);
- [x] modify `IPSocket`, `TCPSocket`, `TCPServer` and `UDPSocket` to use `DNS.getaddrinfo`;
- [x] drop `IPSocket.getaddrinfo` ~~or~~ and avoid a `DNS` namespace to use `Socket::Addrinfo` instead;
- [x] ~~consider reducing duplication of `DNS::Addrinfo` with `Socket::IPAddress` if any (?)~~ the Addrinfo wrapper struct was dropped, because of IP4/IP6 issues and it's not particularly useful, unless we want to open a public API (let's see that later);
- [x] ~~specs !~~ this is a private API, socket specs already test that it works;
- [ ] generate bindings for the `AI_*` and `EAI_*` LibC constants for each supported target.

I was wondering about having different resolvers under `DNS` that let applications use the strategy they want, since none is perfect. For example a blocking resolver that merely calls `getaddrinfo`; a threaded resolver that uses threads to call `getaddrinfo` (ie. this PR, should be the default); a resolver that would link against [c-ares](http://c-ares.haxx.se/) for applications that don't care about libc extensions and want a real async DNS resolver (eg: servers).

Note that this solution may be dropped and no longer be the default when the event-loop will become multithreaded. Limiting the number of concurrent calls to `getaddrinfo` in order to not block all threads (like go does) may be preferable.

refs #2745 #2660

ysbaddaden commented on src/thread/queue.cr:
```diff
@@ -0,0 +1,53 @@
+
+    def initialize
+      @que = Deque(T).new(16)
+      @mutex = Thread::Mutex.new
```
Using a mutex is easy, but maybe not optimal in terms of speed. Maybe using http://llvm.org/releases/3.3/docs/LangRef.html#cmpxchg-instruction and http://llvm.org/releases/3.3/docs/LangRef.html#atomicrmw-instruction for acquiring/releasing a lock could lead to better performance.

ysbaddaden replied:
For example try to lock with `cmpxchg(counter, 1, 0)` and yielding the fiber until it succeeds, then release the lock with `modifyrmw(counter, sub, counter, 1)`. We'd need to differentiate between pushing from the event-loop thread (there is a fiber) and calling from a thread (no fiber) and act accordingly by maybe using a semaphore from the thread instead of yielding.

This may be too complex, thought, and we should just not care until we tackle parallelism.

asterite commented:
Thanks! I think this will be part of the solution that eventually will remain in the standard library (that is, firing a thread and invoking `getaddrinfo` there). I'd still like to wait @waj to check this, as he knows pretty well what works and not when firing a new thread. I remember using fibers there didn't work (eventually crashed or the GC entered an allocation loop), not sure if not using fibers but allocating memory or communicating with other threads is OK though (what this PR does). Hopefully we can review and finish this for 0.19.0 :-)

ysbaddaden commented on src/thread/queue.cr:
```diff
@@ -0,0 +1,53 @@
+      loop do
+        @mutex.synchronize do
+          if @que.empty?
+            raise Error.new("queue is empty") unless blocking
```
When called from the event-loop, this should always non blocking, otherwise the next call will block the event loop until a result a available (bad).

It could also merely return nil, allowing the calling Fiber to yield itself, instead of spending resources raising an exception, catching it and then yielding. Thought there won't be any mean to differentiate between a "would block" and a "popping nil" scenarios.

ysbaddaden commented:
Memory usage is stable over time (on linux glibc) when resolving `localhost` in a tight loop. I have a feeling that LLVM frees local variables immediately, or BOEHM GC does its job correctly. Also, the threads never make any call that will result into accessing the event loop (libevent's base), they merely use C syscalls: wait on a Mutex/ConditionVariable, call `getaddrinfo` and nothing more.

ysbaddaden commented:
I made some modifications (with new commits):

- I dropped the Addrinfo wrapper because I ran into issues with the sockaddr C structs, and it's not required for internal stuffs (except to introduce a lot of useless copying). Maybe it could become a nice public API, but let's see that later and keep it private for now.
- I dropped the DNS namespace, and moved the threaded resolver to `Socket::Addrinfo`. I believe it's a better place for them.
- I made some additional cleanup in the socket classes: harmonized coding styles and variable names, added some missing methods to UNIXServer, along with some basic documentation and a few fixes.

Still a work in progress. I must generate POSIX bindings for the remaining supported targets, and squash things a bit.

sdogruyol commented:
Bump on this 👍

Sija commented:
Bump again.

bararchy commented:
@ysbaddaden @RX14 @straight-shoota 
Should we maybe revisit this idea?

ysbaddaden commented:
No, the ability to configure the resolver in #4236 is better.

bararchy commented:
@ysbaddaden makes sense.
Should we maybe close this one then ?
